File: src/TypeScript/src/tungsten/persistent-association.ts
import { PersistentDeque } from "../finger-tree/core.js";
import { defaultComparator, type Comparator } from "../finger-tree/ordering.js";
import { defaultHashPolicy, type HashPolicy } from "../hamt/hash-policy.js";
import { PersistentHashMap } from "../hamt/persistent-hamt.js";

const stampGap = 1n << 32n;
interface Slot<V> { readonly stamp: bigint; readonly value: V }
interface AssociationEntry<K, V> { readonly stamp: bigint; readonly key: K; readonly value: V }
export interface AssociationRemoveResult<K, V> { readonly association: PersistentAssociation<K, V>; readonly value: V }

/** Persistent insertion-ordered map with Tungsten Association positioning rules. */
export class PersistentAssociation<K, V> implements Iterable<readonly [K, V]> {
    readonly #entries: PersistentDeque<AssociationEntry<K, V>>;
    readonly #index: PersistentHashMap<K, Slot<V>>;
    private constructor(entries: PersistentDeque<AssociationEntry<K, V>>, index: PersistentHashMap<K, Slot<V>>) { this.#entries = entries; this.#index = index; }
    public static empty<K, V>(policy: HashPolicy<K> = defaultHashPolicy<K>()): PersistentAssociation<K, V> { return new PersistentAssociation(PersistentDeque.empty(), PersistentHashMap.empty(policy)); }
    public static from<K, V>(pairs: Iterable<readonly [K, V]>, policy: HashPolicy<K> = defaultHashPolicy<K>()): PersistentAssociation<K, V> { return this.empty<K, V>(policy).setItems(pairs); }
    public get size(): number { return this.#entries.size; }
    public get isEmpty(): boolean { return this.#entries.isEmpty; }
    public get policy(): HashPolicy<K> { return this.#index.policy; }
    public containsKey(key: K): boolean { return this.#index.containsKey(key); }
    public get(key: K): V | undefined { return this.#index.get(key)?.value; }
    public getStoredKey(key: K): K | undefined { return this.#index.getEntry(key)?.key; }
    #pair(entry: AssociationEntry<K, V> | undefined): readonly [K, V] | undefined { return entry === undefined ? undefined : [entry.key, entry.value]; }
    public first(): readonly [K, V] | undefined { return this.#pair(this.#entries.front()); }
    public last(): readonly [K, V] | undefined { return this.#pair(this.#entries.back()); }
    public getAt(index: number): readonly [K, V] | undefined { return this.#pair(this.#entries.get(index)); }
    #indexOfStamp(stamp: bigint): number { let low = 0; let high = this.size; while (low < high) { const middle = (low + high) >>> 1; if (this.#entries.get(middle)!.stamp < stamp) low = middle + 1; else high = middle; } if (low >= this.size || this.#entries.get(low)!.stamp !== stamp) throw new Error("Association stamp is absent from order."); return low; }
    public indexOfKey(key: K): number { const slot = this.#index.getEntry(key)?.value; return slot === undefined ? -1 : this.#indexOfStamp(slot.stamp); }
    public setItem(key: K, value: V): PersistentAssociation<K, V> {
        const found = this.#index.getEntry(key); if (found === undefined) return this.#insertNew(this.#entries, this.#index, this.size, key, value);
        const slot = found.value; if (Object.is(slot.value, value)) return this;
        const position = this.#indexOfStamp(slot.stamp); const stored = this.#entries.get(position)!;
        return new PersistentAssociation(this.#entries.setItem(position, { stamp: slot.stamp, key: stored.key, value })!, this.#index.put(key, { stamp: slot.stamp, value }));
    }
    public setItems(pairs: Iterable<readonly [K, V]>): PersistentAssociation<K, V> { let result: PersistentAssociation<K, V> = this; for (const [key, value] of pairs) result = result.setItem(key, value); return result; }
    public join(other: PersistentAssociation<K, V>): PersistentAssociation<K, V> { if (this.isEmpty && this.policy === other.policy) return other; return this.setItems(other); }
    public append(key: K, value: V): PersistentAssociation<K, V> {
        const found = this.#index.getEntry(key); if (found === undefined) return this.#insertNew(this.#entries, this.#index, this.size, key, value);
        const position = this.#indexOfStamp(found.value.stamp); if (position === this.size - 1 && Object.is(found.value.value, value)) return this;
        const trimmed = this.#entries.removeAt(position)!; return this.#insertNew(trimmed, this.#index.remove(key), trimmed.size, key, value);
    }
    public prepend(key: K, value: V): PersistentAssociation<K, V> {
        const found = this.#index.getEntry(key);
        if (found === undefined) return this.#insertNew(this.#entries, this.#index, 0, key, value);
        const position = this.#indexOfStamp(found.value.stamp); if (position === 0 && Object.is(found.value.value, value)) return this;
        const trimmed = this.#entries.removeAt(position)!; return this.#insertNew(trimmed, this.#index.remove(key), 0, key, value);
    }
    public insert(position: number, key: K, value: V): PersistentAssociation<K, V> | undefined {
        if (!Number.isInteger(position) || position < 0 || position > this.size) return undefined;
        let entries = this.#entries; let index = this.#index; let target = position; const found = index.getEntry(key);
        if (found !== undefined) { const old = this.#indexOfStamp(found.value.stamp); entries = entries.removeAt(old)!; index = index.remove(key); if (old < target) target--; }
        return this.#insertNew(entries, index, target, key, value);
    }
    public remove(key: K): PersistentAssociation<K, V> { return this.tryRemove(key)?.association ?? this; }
    public tryRemove(key: K): AssociationRemoveResult<K, V> | undefined { const removed = this.#index.tryRemove(key); if (removed === undefined) return undefined; const position = this.#indexOfStamp(removed.value.stamp); return { association: new PersistentAssociation(this.#entries.removeAt(position)!, removed.map), value: removed.value.value }; }
    public removeRange(keys: Iterable<K>): PersistentAssociation<K, V> { let result: PersistentAssociation<K, V> = this; for (const key of keys) result = result.remove(key); return result; }
    public keyTake(keys: Iterable<K>): PersistentAssociation<K, V> { let result = PersistentAssociation.empty<K, V>(this.policy); for (const key of keys) { const found = this.#index.getEntry(key); if (found !== undefined && !result.containsKey(key)) result = result.append(found.key, found.value.value); } return result; }
    public removeAt(position: number): PersistentAssociation<K, V> | undefined { const entry = this.#entries.get(position); return entry === undefined ? undefined : new PersistentAssociation(this.#entries.removeAt(position)!, this.#index.remove(entry.key)); }
    public removeFirst(): PersistentAssociation<K, V> | undefined { return this.removeAt(0); }
    public removeLast(): PersistentAssociation<K, V> | undefined { return this.removeAt(this.size - 1); }
    public getRange(position: number, count: number): PersistentAssociation<K, V> | undefined {
        const split = this.#entries.splitRange(position, count); if (split === undefined) return undefined; if (count === this.size) return this;
        let index = PersistentHashMap.empty<K, Slot<V>>(this.policy); for (const entry of split.range) index = index.put(entry.key, { stamp: entry.stamp, value: entry.value });
        return new PersistentAssociation(split.range, index);
    }
    public take(count: number): PersistentAssociation<K, V> | undefined { return this.getRange(0, count); }
    public drop(count: number): PersistentAssociation<K, V> | undefined { return !Number.isInteger(count) || count < 0 || count > this.size ? undefined : this.getRange(count, this.size - count); }
    public reverse(): PersistentAssociation<K, V> { return this.size <= 1 ? this : this.#rebuild(this.#entries.toArray().reverse()); }
    public keySort(comparator: Comparator<K> = defaultComparator): PersistentAssociation<K, V> { return this.size <= 1 ? this : this.#rebuild(this.#entries.toArray().sort((left, right) => comparator(left.key, right.key) || Number(left.stamp - right.stamp))); }
    public sort(comparator: Comparator<V> = defaultComparator): PersistentAssociation<K, V> { return this.size <= 1 ? this : this.#rebuild(this.#entries.toArray().sort((left, right) => comparator(left.value, right.value) || Number(left.stamp - right.stamp))); }
    public keys(): K[] { return this.#entries.toArray().map((entry) => entry.key); }
    public values(): V[] { return this.#entries.toArray().map((entry) => entry.value); }
    public toArray(): Array<readonly [K, V]> { return this.#entries.toArray().map((entry) => [entry.key, entry.value]); }
    #pickStamp(entries: PersistentDeque<AssociationEntry<K, V>>, position: number): bigint | undefined { if (entries.isEmpty) return 0n; if (position === 0) return entries.front()!.stamp - stampGap; if (position === entries.size) return entries.back()!.stamp + stampGap; const left = entries.get(position - 1)!.stamp; const right = entries.get(position)!.stamp; return right - left <= 1n ? undefined : left + (right - left) / 2n; }
    #insertNew(entries: PersistentDeque<AssociationEntry<K, V>>, index: PersistentHashMap<K, Slot<V>>, position: number, key: K, value: V): PersistentAssociation<K, V> { const stamp = this.#pickStamp(entries, position); if (stamp === undefined) { const ordered = entries.toArray(); ordered.splice(position, 0, { stamp: 0n, key, value }); return this.#rebuild(ordered); } const entry = { stamp, key, value }; const nextEntries = position === 0 ? entries.prepend(entry) : position === entries.size ? entries.append(entry) : entries.insertAt(position, entry)!; return new PersistentAssociation(nextEntries, index.put(key, { stamp, value })); }
    #rebuild(ordered: readonly AssociationEntry<K, V>[]): PersistentAssociation<K, V> { let index = PersistentHashMap.empty<K, Slot<V>>(this.policy); const entries = ordered.map((entry, position) => { const stamp = BigInt(position) * stampGap; index = index.put(entry.key, { stamp, value: entry.value }); return { stamp, key: entry.key, value: entry.value }; }); return new PersistentAssociation(PersistentDeque.from(entries), index); }
    public *[Symbol.iterator](): IterableIterator<readonly [K, V]> { for (const entry of this.#entries) yield [entry.key, entry.value]; }
}

File: src/OCaml/lib/tungsten/persistent_association.ml
type ('key, 'value) entry = { key : 'key; value : 'value }

type ('key, 'value) t = {
  equality : 'key Common.Hash_policy.t;
  value_equal : 'value -> 'value -> bool;
  entries : ('key, 'value) entry array;
}

let empty ?(value_equal = ( = )) equality = { equality; value_equal; entries = [||] }
let policy association = association.equality
let count association = Array.length association.entries
let is_empty association = count association = 0
let entry_key entry = entry.key
let entry_value entry = entry.value
let equal association = Common.Hash_policy.equal association.equality

let index_of key association =
  let rec search index =
    if index = count association then None
    else if equal association association.entries.(index).key key then Some index
    else search (index + 1)
  in
  search 0

let mem key association = Option.is_some (index_of key association)

let find_entry key association =
  Option.map (fun index -> association.entries.(index)) (index_of key association)

let find_opt key association = Option.map entry_value (find_entry key association)

let nth index association =
  if index < 0 || index >= count association then None else Some association.entries.(index)

let first association = nth 0 association
let last association = nth (count association - 1) association

let set key value association =
  match index_of key association with
  | None ->
      {
        association with
        entries =
          Finger_tree.Sorted_helpers.insert (count association) { key; value } association.entries;
      }
  | Some index when association.value_equal association.entries.(index).value value -> association
  | Some index ->
      let entries = Array.copy association.entries in
      entries.(index) <- { key = entries.(index).key; value };
      { association with entries }

let set_items pairs association =
  List.fold_left (fun result (key, value) -> set key value result) association pairs

let of_list ?(value_equal = ( = )) equality pairs = set_items pairs (empty ~value_equal equality)

let to_list association =
  Array.to_list (Array.map (fun entry -> (entry.key, entry.value)) association.entries)

let join left right = set_items (to_list right) left

let insert position key value association =
  if position < 0 || position > count association then
    Error "Tungsten Association insertion position is out of bounds"
  else
    let entries, target =
      match index_of key association with
      | None -> (association.entries, position)
      | Some previous ->
          ( Finger_tree.Sorted_helpers.remove previous association.entries,
            if previous < position then position - 1 else position )
    in
    Ok
      { association with entries = Finger_tree.Sorted_helpers.insert target { key; value } entries }

let append key value association = Result.get_ok (insert (count association) key value association)
let prepend key value association = Result.get_ok (insert 0 key value association)

let remove key association =
  Option.map
    (fun index ->
      ( association.entries.(index).value,
        { association with entries = Finger_tree.Sorted_helpers.remove index association.entries }
      ))
    (index_of key association)

let remove_at index association =
  if index < 0 || index >= count association then
    Error "Tungsten Association removal position is out of bounds"
  else Ok { association with entries = Finger_tree.Sorted_helpers.remove index association.entries }

let key_take requested association =
  List.fold_left
    (fun result key ->
      match (find_entry key association, find_entry key result) with
      | Some entry, None -> append entry.key entry.value result
      | Some _, Some _ | None, _ -> result)
    (empty ~value_equal:association.value_equal association.equality)
    requested

let range ~start ~count:range_count association =
  if start < 0 || range_count < 0 || start > count association - range_count then
    Error "Tungsten Association range is out of bounds"
  else
    Ok
      {
        association with
        entries = Finger_tree.Sorted_helpers.slice start range_count association.entries;
      }

let reverse association =
  { association with entries = Array.of_list (List.rev (Array.to_list association.entries)) }

let sort_by project comparator association =
  let entries = Array.copy association.entries in
  Array.stable_sort
    (fun left right -> Common.Comparator.compare comparator (project left) (project right))
    entries;
  { association with entries }

let key_sort comparator = sort_by (fun entry -> entry.key) comparator
let value_sort comparator = sort_by (fun entry -> entry.value) comparator
let keys association = List.map fst (to_list association)
let values association = List.map snd (to_list association)
